include/dd/NoiseFunctionality.hpp

```diff
@@ -0,0 +1,714 @@
+                } else {
+                    tmp.p = originalEdge.p;
+                    return tmp;
+                }
+            }
+            auto originalCopy = originalEdge;
+            originalCopy.w    = dd::Complex::one;
+
+            // Check if the target of the current edge is in the "compute table".
+
+            //    auto noiseLookUpResult = dd->densityNoise.lookup(originalCopy, usedQubits);
+            //
+            //    if (noiseLookUpResult.p != nullptr) {
+            //        auto tmpComplexValue = dd->cn.getCached();
+            //        CN::mul(tmpComplexValue, noiseLookUpResult.w, originalEdge.w);
+            //        noiseLookUpResult.w = dd->cn.lookup(tmpComplexValue);
+            //        dd->cn.returnToCache(tmpComplexValue);
+            //        return noiseLookUpResult;
+            //    }
+            std::array<dEdge, std::tuple_size_v<decltype(dd::dNode::e)>> new_edges{};
+            for (short i = 0; i < 4; i++) {
+                if (firstPathEdge || i == 1) {
+                    // If I am to the useDensityMatrix I cannot minimize the necessary operations anymore
+                    dEdge::applyDmChangesToEdge(&originalCopy.p->e[i]);
+                    new_edges[i] = applyNoiseEffects(originalCopy.p->e[i], usedQubits, true);
+                    dEdge::revertDmChangesToEdge(&originalCopy.p->e[i]);
+                } else if (i == 2) {
+                    // Size e[1] == e[2] (due to density matrix representation), I can skip calculating e[2]
+                    new_edges[2].p = new_edges[1].p;
+                    new_edges[2].w = package->cn.getCached(CTEntry::val(new_edges[1].w.r), CTEntry::val(new_edges[1].w.i));
+                } else {
+                    dEdge::applyDmChangesToEdge(&originalCopy.p->e[i]);
+                    new_edges[i] = applyNoiseEffects(originalCopy.p->e[i], usedQubits, false);
+                    dEdge::revertDmChangesToEdge(&originalCopy.p->e[i]);
+                }
+            }
+            dEdge e = {};
+            if (std::count(usedQubits.begin(), usedQubits.end(), originalCopy.p->v)) {
+                for (auto const& type: noiseEffects) {
+                    switch (type) {
+                        case dd::amplitudeDamping:
+                            applyAmplitudeDampingToEdges(new_edges, (usedQubits.size() == 1) ? ampDampingProbSingleQubit : ampDampingProbMultiQubit);
+                            break;
+                        case dd::phaseFlip:
+                            applyPhaseFlipToEdges(new_edges, (usedQubits.size() == 1) ? noiseProbSingleQubit : noiseProbMultiQubit);
+                            break;
+                        case dd::depolarization:
+                            applyDepolarisationToEdges(new_edges, (usedQubits.size() == 1) ? noiseProbSingleQubit : noiseProbMultiQubit);
+                            break;
+                        case dd::identity:
+                            continue;
```

Comment:
it should be easy to cover these lines with tests

Reply:
done